lib/xgu/xgu.h

```diff
@@ -0,0 +1,993 @@
+    return push_command_boolean(p, NV097_SET_DITHER_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_lighting_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_LIGHTING_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_stencil_test_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_STENCIL_TEST_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_alpha_func(uint32_t* p, XguFuncType func) {
+    return push_command_parameter(p, NV097_SET_ALPHA_FUNC, func);
+}
+
+inline
+uint32_t* xgu_set_alpha_ref(uint32_t* p, uint8_t ref) {
+    return push_command_parameter(p, NV097_SET_ALPHA_REF, ref);
+}
+
+inline
+uint32_t* xgu_set_blend_func_sfactor(uint32_t* p, XguBlendFactor sf) {
+    return push_command_parameter(p, NV097_SET_BLEND_FUNC_SFACTOR, sf);
+}
+
+inline
+uint32_t* xgu_set_blend_func_dfactor(uint32_t* p, XguBlendFactor df) {
+    return push_command_parameter(p, NV097_SET_BLEND_FUNC_DFACTOR, df);
+}
+
+inline
+uint32_t* xgu_set_depth_func(uint32_t* p, XguFuncType func) {
+    return push_command_parameter(p, NV097_SET_DEPTH_FUNC, func);
+}
+
+inline
+uint32_t* xgu_set_color_mask(uint32_t* p, XguColorMask cm) {
+    return push_command_parameter(p, NV097_SET_COLOR_MASK, cm);
+}
+
+inline
+uint32_t* xgu_set_front_polygon_mode(uint32_t* p, XguPolygonMode pm) {
+    return push_command_parameter(p, NV097_SET_FRONT_POLYGON_MODE, pm);
+}
+
+inline
+uint32_t* xgu_set_cull_face(uint32_t* p, XguCullFace cf) {
+    return push_command_parameter(p, NV097_SET_CULL_FACE, cf);
+}
+
+inline
+uint32_t* xgu_set_front_face(uint32_t* p, XguFrontFace ff) {
+    return push_command_parameter(p, NV097_SET_FRONT_FACE, ff);
+}
+
+/* ==== Transform ==== */
+inline
+uint32_t* xgu_set_transform_execution_mode(uint32_t* p, XguExecMode mode, XguExecRange range) {
+    return push_command_parameter(p,
+                                  NV097_SET_TRANSFORM_EXECUTION_MODE,
+                                  XGU_MASK(NV097_SET_TRANSFORM_EXECUTION_MODE_MODE, mode) |
+                                  XGU_MASK(NV097_SET_TRANSFORM_EXECUTION_MODE_RANGE_MODE, range));
+}
+
+inline
+uint32_t* xgu_set_transform_constant(uint32_t* p, const XguVec4 *v, unsigned int count) {
+    p = push_command(p, NV097_SET_TRANSFORM_CONSTANT, count*4);
+    for (uint32_t i = 0; i < count; ++i) {
+        p = push_floats(p, v[i].f, 4);
+    }
+    return p;
+}
+
+inline
+uint32_t* xgu_set_transform_constant_load(uint32_t* p, uint32_t offset) {
+    return push_command_parameter(p, NV097_SET_TRANSFORM_CONSTANT_LOAD, offset);
+}
+
+
+typedef union {
+    uint32_t i[4];
+    struct {
+        uint32_t a;
+        uint32_t b;
+        uint32_t c;
+        uint32_t d;
+    };
+} XguTransformProgramInstruction;
+
+inline
+uint32_t* xgu_set_transform_program(uint32_t* p, const XguTransformProgramInstruction* instructions, unsigned int count) {
+    p = push_command(p, NV097_SET_TRANSFORM_PROGRAM, count*4);
```

Comment:
This is broken: https://discord.com/channels/428359196719972353/428399682239332354/708473483386028040

(Only supports up to 8 instructions at the same time in incremental method mode. Not sure if non-incremental mode is supported)